### PR TITLE
feat: ChangeTierRequest updates Space

### PR DESCRIPTION
**Pull request by xcoulon**
Apply the ChangeTierRequest on a Space with the same
name as the target MasterUserRecord, if it exists.
Only care about the `tierName` part of the ChangeTierRequest,
since the Space resource already caries the `targetCluster` info.

Also, rename all occurrences of `space` var to `s` to avoid conflict
with the `github.com/codeready-toolchain/host-operator/controllers/space`
pkg.

See also: 
- https://github.com/codeready-toolchain/api/pull/278
- https://github.com/codeready-toolchain/toolchain-common/pull/221
- https://github.com/codeready-toolchain/toolchain-e2e/pull/420

Fixes https://issues.redhat.com/browse/CRT-1311

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

**Comment by xcoulon** [comment]
in draft for now: will rebase this PR once #548 is merged

File: controllers/space/space_controller.go
Comment (rajivnathan) [inline controllers/space/space_controller.go:142]:
```suggestion
				return false, err
```
If `r.setStatusProvisioning(space)` then we should probably just return false, err instead?

File: controllers/space/space_controller.go
Comment (rajivnathan) [inline controllers/space/space_controller.go:366]: I'm a bit confused by the name `setStatusChangeFailed`. Since it is setting the ready condition to false with reason `SpaceStatusUpdateFailedReason` then would it be better to rename it to `setStatusUpdateFailed`

File: controllers/space/space_controller.go
Comment (rajivnathan) [inline controllers/space/space_controller.go:375]: I guess this is not the right error to log?

File: controllers/changetierrequest/changetierrequest_controller.go
Comment (MatousJobanek) [reply controllers/changetierrequest/changetierrequest_controller.go:243]: Well, this is changed in one update call, so both of the values are updated together - it should not happen that the tierName is changed but the label stays there. Correct me if I'm wrong.
A different story is a case when the `ChangeTierRequest` doesn't change the `space.Spec.TierName` at all:
1. `Space.Spec.TierName=advanced`
2. `ChangeTierRequest.Spec.TierName=advanced` 
3. Then the `ChangeTierRequest` controller only deletes the label but doesn't change the spec. However, the `Space` controller won't be triggered since it doesn't watch the changes of the labels, so it won't set the label back.
There should be something like:
```go
if space.Spec.TierName == changeTierRequest.Spec.TierName {
	return nil
}
space.Spec.TierName = changeTierRequest.Spec.TierName
// remove the TemplateTierHash label on the Space resource (and let the SpaceController set it to the latest value)
delete(space.Labels, tierutil.TemplateTierHashLabelKey(space.Spec.TierName))
```

File: controllers/space/space_controller_test.go
Comment (MatousJobanek) [inline controllers/space/space_controller_test.go:650]: it's the same thing as last time - we don't need to create the reconcile again when the client and controller is supposed to be reused in the sub-test
```suggestion
```

File: controllers/space/space_controller_test.go
Comment (MatousJobanek) [inline controllers/space/space_controller_test.go:673]:
```suggestion
```

File: controllers/space/space_controller_test.go
Comment (MatousJobanek) [inline controllers/space/space_controller_test.go:800]:
```suggestion
```

File: controllers/space/space_controller_test.go
Comment (MatousJobanek) [inline controllers/space/space_controller_test.go:823]:
```suggestion
```

File: controllers/space/space_controller_test.go
Comment (MatousJobanek) [inline controllers/space/space_controller_test.go:846]:
```suggestion
```

File: test/space/space.go
Comment (MatousJobanek) [inline test/space/space.go:30]: it looks like that it's not used anywhere

File: test/space/space_assertions.go
Comment (MatousJobanek) [inline test/space/space_assertions.go:158]: probably not used either

File: test/space/space.go
Comment (xcoulon) [reply test/space/space.go]: removed in 9e9d8ec

File: test/space/space_assertions.go
Comment (xcoulon) [reply test/space/space_assertions.go]: removed in 9e9d8ec

File: controllers/changetierrequest/changetierrequest_controller_test.go
Comment (xcoulon) [reply controllers/changetierrequest/changetierrequest_controller_test.go]: other tests covered in 1eca554

File: controllers/changetierrequest/changetierrequest_controller.go
Comment (xcoulon) [reply controllers/changetierrequest/changetierrequest_controller.go]: done in 56896f0

File: controllers/changetierrequest/changetierrequest_controller.go
Comment (xcoulon) [reply controllers/changetierrequest/changetierrequest_controller.go:265]: done in 2f3805e

File: controllers/changetierrequest/changetierrequest_controller_test.go
Comment (xcoulon) [reply controllers/changetierrequest/changetierrequest_controller_test.go:187]: done in 70e0313

File: controllers/changetierrequest/changetierrequest_controller_test.go
Comment (xcoulon) [reply controllers/changetierrequest/changetierrequest_controller_test.go]: done in e431ecc, dropping the case where MUR exists or not in these particular subtests

File: controllers/space/space_controller.go
Comment (xcoulon) [reply controllers/space/space_controller.go:142]: this code is no outdated, because the reason for failing was when fetching the NSTemplateTier would fail. This block has been extracted into:
```
tmplTier := &toolchainv1alpha1.NSTemplateTier{}
if err := r.Client.Get(context.TODO(), types.NamespacedName{
	Namespace: space.Namespace,
	Name:      space.Spec.TierName,
}, tmplTier); err != nil {
	return false, r.setStatusProvisioningFailed(logger, space, err)
}
```
but I believe that it's valid to set the status of the Space to `UnableToProvision` if the Tier does not exist, isn't it?

File: controllers/space/space_controller.go
Comment (xcoulon) [reply controllers/space/space_controller.go]: this code doesn't exist anymore

File: controllers/space/space_controller.go
Comment (xcoulon) [reply controllers/space/space_controller.go]: this block of code was removed

File: controllers/space/space_controller.go
Comment (xcoulon) [reply controllers/space/space_controller.go:158]: ok, so I will leave it as-is for now (too many other comments to take into account for now! 🥵)

File: controllers/space/space_controller.go
Comment (xcoulon) [reply controllers/space/space_controller.go:202]: I'm sorry but I don't understand why we should set the Space status to `ProvisioningFailed` while the NSTemplateSet is updating. It's not a case of failure, is it?

File: controllers/space/space_controller.go
Comment (xcoulon) [reply controllers/space/space_controller.go]: updated in f4b8401 after discussing/clarifying on Google Meet

File: controllers/space/space_controller.go
Comment (xcoulon) [reply controllers/space/space_controller.go]: sure, so should this `ensureNSTemplateSet` method only return `true` when the NSTemplateSet was updated, and false in all other cases (created, unchanged or when a failure occurred)?

File: controllers/space/space_controller.go
Comment (xcoulon) [reply controllers/space/space_controller.go:223]: so, to clarify: are you ok to keep the code as I proposed?

File: controllers/space/space_controller.go
Comment (rajivnathan) [reply controllers/space/space_controller.go:159]: What I was trying to say before was that if `r.setStatusProvisioning(space)` was returning an error then trying to do `r.setStatusProvisioningFailed` wouldn't make much sense because it's another status update. ie. if one status update failed, we probably shouldn't try to make another status update

But it's changed now so what you have is fine. 👍

**Comment by rajivnathan** [comment]
Thanks @xcoulon for addressing my comments! 🙌  I have no further concerns but I'll leave the final approval for Matous.

File: controllers/space/space_controller.go
Comment (MatousJobanek) [reply controllers/space/space_controller.go:223]: The thing that I don't like about this switch-case solution is the level of complexity. I spent some time reading the code but I really don't understand when it sets what condition. That's why I proposed the simple one `if` statement which is much easier to read & understand.

Comment (xcoulon) [reply controllers/space/space_controller.go]: @MatousJobanek I simplified the logic of translating the NSTemplateSet status into the Space in cd0ad5c based on your feedback and our discussion and IMO it's much easier to read now. Please let me know what you think about it ;)

File: controllers/space/space_controller.go
Comment (MatousJobanek) [inline controllers/space/space_controller.go:185]:
```suggestion
```

File: controllers/space/space_controller.go
Comment (MatousJobanek) [inline controllers/space/space_controller.go:211]: btw, do we need two immediate updates of the same Space? Couldn't that be done/updated together?

**Review by MatousJobanek — COMMENTED** [review]
The update of the condition looks much better - thanks a lot :+1: 
However, the requeue is still executed/returned at the wrong moment.

File: controllers/space/space_controller.go
Comment (xcoulon) [reply controllers/space/space_controller.go:211]: there's an update of the labels and an update of the status. Last time I ran the e2e tests, I needed to make those 2 calls explicitly. But I can check again to see if I didn't miss anything.

Comment (MatousJobanek) [reply controllers/space/space_controller.go:210]: if you just set the label and executed the `setStatusProvisioned` method then it makes sense - the `updateStatus` method doesn't call Update if there is no change in the condition.

File: controllers/space/space_controller.go
Comment (xcoulon) [reply controllers/space/space_controller.go:135]: yes, I remember about this discussion and I thought I already addressed it 😲

Comment (xcoulon) [reply controllers/space/space_controller.go]: addressed in fb396b3

File: controllers/space/space_controller.go
Comment (xcoulon) [reply controllers/space/space_controller.go]: hopefully addressed in 6b76c33 after our discussion on Google Meet earlier today

**Review by MatousJobanek — APPROVED** [review]
Great work :+1:

**Pull request merged**
